draft/XXXX-robust-api/README.md

```diff
@@ -0,0 +1,146 @@
+
+### Upload Collection
+
+Each workspace should contain its own upload collection, which will be used to store each upload document shown above. This will need to be accompanied by a new `Upload` ORM, which will facilitate operations on uploads.
+
+### Additional Upload Endpoints
+
+`GET /api/workspaces/{workspace}/upload/{id}`
+
+This endpoint returns the Upload document with the corresponding ID.
+
+## Open Questions
+
+### Changes to data processing procedure
+
+The notion of having various file type "uploaders" will be removed. Instead, these will become data processors, which will take raw uploaded data (from S3), and run validation, table/graph creation, data insertion, etc. When downloading the finalized upload from S3, this processing should be done in a streaming manner, in order to limit the risk of running out of memory in the server. This will involve updating each data processor to handle streamed data, instead of the entire file. This could be done in the following manner:
+
+1. Read in data until a fully formed object can be parsed (a table row, for example).
+2. Run validation on this object
+3. If validation is passed, insert this object into the target location. If validation fails, save the validation error, and either continue processing more data, or halt data processing.
+4. Repeat steps 1-3 until all data has been processed
+5. If there are errors resulting from this processing, delete the partially formed tables/graphs, and return the validation errors to the user.
+
+These errors would likely be stored in the upload model (yet to be outlined), however, it may make sense to introduce another model that would track this status more specifically. Feedback is welcome here.
+
+
+### Tracking upload progress
+
+The current proposal requires that the client upload all parts, and then returns the ETags of these parts all at once when finalizing the upload. However, this poses a couple of possible pitfalls:
+
+1. There is no tracking of upload progress on the server
+2. If it's desired to allow for upload pausing and resuming (assuming the page is totally refreshed and all context is lost, or the computer is changed, etc.), there is no way to do this, as the initially acquired ETags are now lost
+
+This could be addressed by creating endpoints for finalizing an individual part upload, as well as the overall finalize call. In this case, the procedure would then require the client to call some endpoint with the following signature:
+
+```
+POST /api/workspaces/{workspace}/upload/{id}/finalizePart
+
+{
+  "partNumber": integer,
+  "ETag": string
+}
+```
+
+This would add to the `uploadPartTags` array on the upload document, which would allow for tracking the progress of this upload. After all file parts are uploaded, the upload finalize call is made as usual, but in this new approach, it would not contain any payload.
+
+
+Feedback on this suggested feature is appreciated.
```

Comment:
The simpler approach is to have the client save the ETags in local storage. That would cover the common case, which is that an upload breaks in the user's browser, and the user fixes it from the same browser.

Recording successful part uploads by informing the server of each one is a more robust solution, but seems like it will require further design. It is an enhancement to the current proposal, so TBH I would say don't worry about it right now, and defer work on resumable uploads in a future push.

Reply:
There's two ways I can see this, as far as simplicity.:

1. The original proposal is simpler, because you're sending less requests back to the server, and storing things locally. However, that does mean you're storing two versions of the upload progress between the client/server.
2. The second approach is simpler, because it consolidates storage into the server, and the client isn't responsible for keeping track of everything, just doing its part to return it's answer to the server. However, that does involve more (albeit lightweight) requests.

You could go either way on these, but as far as API design, I'm more in favor of the second option. This is because if you _were_ to go with the first option, the client would be "keeping" information from the server. That is, if (for example) the browser is closed halfway through the upload, the API has no idea how far along the upload is, just that it's been started, and not finished. If that local storage was cleared for some reason, the client would have to re-upload all of those uploaded parts again. I know this is fairly unlikely to happen, but some people care about what information websites store in their browser etc., and would maybe delete this somehow. If we store this information on the server, it's nobody's burden to track except our own.

Let me know what you think.

Reply:
I accept your argument.

My approach is only "simpler" in the sense that it requires less server code. It still requires (client) code, and it has the other non-simplicities you mentioned.